fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/PushDownNotSlotReferenceExpressionOfOnClause.java

```diff
@@ -0,0 +1,200 @@
+// under the License.
+
+package org.apache.doris.nereids.rules.rewrite.logical;
+
+import org.apache.doris.nereids.rules.Rule;
+import org.apache.doris.nereids.rules.RuleType;
+import org.apache.doris.nereids.rules.rewrite.OneRewriteRuleFactory;
+import org.apache.doris.nereids.trees.expressions.Alias;
+import org.apache.doris.nereids.trees.expressions.Arithmetic;
+import org.apache.doris.nereids.trees.expressions.ComparisonPredicate;
+import org.apache.doris.nereids.trees.expressions.CompoundPredicate;
+import org.apache.doris.nereids.trees.expressions.Expression;
+import org.apache.doris.nereids.trees.expressions.Literal;
+import org.apache.doris.nereids.trees.expressions.NamedExpression;
+import org.apache.doris.nereids.trees.expressions.SlotReference;
+import org.apache.doris.nereids.trees.expressions.visitor.ExpressionVisitor;
+import org.apache.doris.nereids.trees.plans.GroupPlan;
+import org.apache.doris.nereids.trees.plans.Plan;
+import org.apache.doris.nereids.trees.plans.logical.LogicalJoin;
+import org.apache.doris.nereids.trees.plans.logical.LogicalProject;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Lists;
+
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * push down expression which is not slot reference
+ */
+public class PushDownNotSlotReferenceExpressionOfOnClause extends OneRewriteRuleFactory {
+
+    private static class ListBuilder<T> {
+        List<T> list;
+
+        public ListBuilder(List<T> list) {
+            this.list = Lists.newArrayList(list);
+        }
+
+        public ListBuilder<T> addAll(List<T> element) {
+            list.addAll(Lists.newArrayList(element));
+            return this;
+        }
+
+        public List<T> get() {
+            return list;
+        }
+    }
+
+    private final Map<SlotReference, Map<Expression, Expression>> exprMap = new HashMap<>();
+
+    /**
+     * rewrite example:
+     *join(t1.a + 1 = t2.b + 2 and t1.a > 2)        join(c = d and c > 2)
+     *          │                                             │
+     *          ├─olapScan(t1)                                ├────project(t1.a + 1 as c)
+     *          │                               ====>         │       │
+     *          └─olapScan(t2)                                │       └─olapScan(t1)
+     *                                                        │
+     *                                                        └────project(t2.b + 2 as d)
+     *                                                                │
+     *                                                                └─olapScan(t2)
+     */
+    @Override
+    public Rule build() {
+        return logicalJoin()
+                .when(join -> join.getCondition().isPresent()
```

Comment:
This rule is not applicable to all join types. for example, we cannot push on predicate of left outer join down.
So we have to take join type into consideration.

Reply:
Yes, I will change it, we support only inner join first, then I rethink about the type of the join supported.